aws/sts.go

```diff
@@ -2,25 +2,46 @@ package aws
 
 import (
 	"fmt"
+	"strings"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/fatih/color"
 )
 
-func AssumeSAMLRole(PrincipalArn, RoleArn, SAMLAssertion string) (*Credentials, error) {
+// AssumeSAMLRole asumes a Role using the SAMLAssertion specified. If the duration cannot be meet it transperently lowers the duration and sets the returned bool to true to signal that a message should be displayed
+func AssumeSAMLRole(PrincipalArn, RoleArn, SAMLAssertion, profile string, duration int64) (*Credentials, error) {
+	creds, err, roleNeedsChange := assumeSAMLRole(PrincipalArn, RoleArn, SAMLAssertion, duration, false)
+	if err == nil && roleNeedsChange {
+		// TODO: This might conflict with spinner... Trying to avoid by starting with \r.
+		fmt.Printf(color.YellowString("\rThe role does not support the requested max-session-duration of %v. To have a max session duration for up to 12h run:\n"), duration)
+		fmt.Printf("\raws iam update-role --role-name %v --max-session-duration 43200 --profile %v\n", RoleArn[strings.LastIndex(RoleArn, "/")+1:], profile)
+	}
+	return creds, err
+}
+
+func assumeSAMLRole(PrincipalArn, RoleArn, SAMLAssertion string, duration int64, roleDoesNotSupportDuration bool) (*Credentials, error, bool) {
+	// Assume role
 	input := sts.AssumeRoleWithSAMLInput{
-		PrincipalArn:  aws.String(PrincipalArn),
-		RoleArn:       aws.String(RoleArn),
-		SAMLAssertion: aws.String(SAMLAssertion),
+		PrincipalArn:    aws.String(PrincipalArn),
+		RoleArn:         aws.String(RoleArn),
+		SAMLAssertion:   aws.String(SAMLAssertion),
+		DurationSeconds: aws.Int64(duration),
 	}
 
 	sess := session.Must(session.NewSession())
 	svc := sts.New(sess)
 
 	aResp, err := svc.AssumeRoleWithSAML(&input)
 	if err != nil {
-		return nil, fmt.Errorf("assuming role: %v", err)
+		// The role might not yet support the requested duration, let's catch and try to lower in 1h steps
+		if strings.HasPrefix(err.Error(), "ValidationError: The requested DurationSeconds exceeds the MaxSessionDuration set for this role") && duration > 3600 && duration <= 43200 {
```

Comment:
Unfortunately there is no other way. The error code is generic. For the why see above comment.

Reply:
So the purpose is the following: By default I have set my provider to work with 8 hours sessions. So they will last a around workday. But I often work with profiles where this has not been set. So without the code above there are two options:

1. Fix the role before adding it (i.e. open AWS, alter the role and then add it to Clisso).
2. Add the role with 1 hour, get creds using Clisso, alter the role using the cli and then update the config.

Both options suck. So I implemented the third.

3. Try what the user wants and then lower in 1h steps if that's not possible. Because some accounts might have a different requirement or are not setup yet. So a default of e.g. 12h will get you max possible duration on all accounts. And only fails if it can't even get a 1h session.

Reply:
Thanks for moving the error handling further up the stack. It already looks better and I think the spinner problems are gone (right?).

I understand where you're coming from, I really do. But we have to think about the quality of Clisso and the user experience we're providing if we want more people to start using the tool.

We have to remember that Clisso is just a **user interface** to authenticate to AWS. It is 100% dependent on the configuration on OneLogin and AWS, and cannot do anything that is not supported by these two platforms. The problem I have with the approach you're suggesting (even if it works and we make the code beautiful), is that (in my opinion) we are **exceeding Clisso's boundaries** here. The best we can do when Clisso's config doesn't match something on OneLogin or AWS is alert the user about that so that they can fix it. But here we are trying to compensate for wrong config (Clisso's duration > allowed role duration) which is outside the scope of Clisso's responsibility. Not only that, here is the price we're paying for this:

- We bother the user one something they might not care about. For example I personally always work with 1h sessions and actually prefer it this way for security reasons.
- We make the code more complicated.

I'm not suggesting we ditch the custom session duration feature, on the contrary - I think it's good and even necessary. What I **do** suggest is the following:

**By default, let's use something that will always work** (I think 1h sessions are the minimum which is always supported?). If the user doesn't know / doesn't care about session duration, or if they are OK with the default, we leave it at 1h **unless the user decides to override this**. This is also preferred for **security** reasons - we request the safest credentials (1h) unless the user explicitly requests longer sessions.
Now, if the user wants longer sessions - that's perfectly fine and we support that. They can specify the session duration at the provider / app level. **However**, we state very clearly in the docs that they must ensure the max session duration on AWS supports the configured value.
When the user runs `clisso get`, if they request a longer session than what is supported, **we exit with an error**, same as we do when any other configuration param is wrong. In this error we can provide useful information regarding the problem, including a link to the docs and even the AWS CLI command to increase the session's duration.

Lastly, even if we do decide it makes sense to be ultra-proactive and compensate for the wrong config in the code, I would try to **query AWS** for the actual supported maximum, lower the requested duration to that value and alert the user to fix their config on Clisso / AWS.

When you think about the above, I want to ask you to think about the user experience we provide to community users who try Clisso, and not only on your specific use case (8h workday). We should aim to provide sane defaults along with "advanced" customization options, and above all - the default behavior should be something that always works and behaves simply and nicely out of the box.

Would love to hear your thoughts.

Reply:
Here is a snippet to query AWS for the desired session duration:

```go
package main

import (
	"fmt"

	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/iam"
)

func main() {
	svc := iam.New(session.New())
	in := iam.ListRolesInput{}

	out, err := svc.ListRoles(&in)
	if err != nil {
		panic(err)
	}
	fmt.Printf("%v\n", *out.Roles[0].MaxSessionDuration)
}
```

If we use that, we don't have to recurse and we don't have to check the error type we got from IAM.

Reply:
Ok, so you already found the chicken-egg problem with querying AWS and the need for credentials. Additionally you'd assume the user can use IAM to list details on the role.

I'm okay with removing the recursion. It means to configure more on the Clisso side if you need different durations and not just changing the limits on AWS. Depending on the angle it might security more hard on the user, but I'd be ok with that.

I'm against failing hard. Because then you don't have any credentials to change the value using the cli. And it doesn't mean they have a an actual wrong configuration. Could be a change on AWS side by someone else.

I'll change the code accordingly.